dali/operators/image/resize/resize_attr.cc

```diff
@@ -0,0 +1,318 @@
+  requested_size.resize(spatial_ndim_);
+  in_size.resize(spatial_ndim_);
+
+  if (HasSeparateSizeArgs()) {
+    res_x_.resize(N);
+    res_y_.resize(N);
+    res_z_.resize(N);
+
+    float *size_vecs[3] = { nullptr, nullptr, nullptr };
+    {
+      // outer-first shape order :(
+      int d = 0;
+      if (spatial_ndim_ >= 3)
+        size_vecs[d++] = res_z_.data();
+      if (spatial_ndim_ >= 2)
+        size_vecs[d++] = res_y_.data();
+      size_vecs[d++] = res_x_.data();
+    }
+
+    if (has_resize_x_) {
+      GetPerSampleArgument(res_x_, "resize_x", spec, ws, N);
+    }
+
+    if (has_resize_y_) {
+      GetPerSampleArgument(res_y_, "resize_y", spec, ws, N);
+    }
+
+    if (has_resize_z_) {
+      GetPerSampleArgument(res_z_, "resize_z", spec, ws, N);
+    }
+
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = input_shape.tensor_shape_span(i)[d + first_spatial_dim_];
+        requested_size[d] = size_vecs[d][i];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size);
+    }
+  } else if (has_resize_shorter_ || has_resize_longer_) {
+    const char *arg_name = has_resize_shorter_ ? "resize_shorter" : "resize_longer";
+    GetPerSampleArgument(res_x_, arg_name, spec, ws, N);
+
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = input_shape.tensor_shape_span(i)[d + first_spatial_dim_];
+        requested_size[d] = res_x_[i];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size);
+    }
+  } else if (has_size_) {
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = input_shape.tensor_shape_span(i)[d + first_spatial_dim_];
+        requested_size[d] = size_arg_.tensor_shape_span(i)[d];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size);
+    }
+  }
+}
+
+void ResizeAttr::AdjustOutputSize(float *out_size, const float *in_size) {
+  SmallVector<double, 3> scale;
+  SmallVector<bool, 3> mask;
+  scale.resize(spatial_ndim_, 1);
+  mask.resize(spatial_ndim_);
+
+  int sizes_provided = 0;
+  for (int d = 0; d < spatial_ndim_; d++) {
+    mask[d] = (out_size[d] != 0);
+    scale[d] = in_size[d] ? out_size[d] / in_size[d] : 1;
+    sizes_provided += mask[d];
+  }
+
+  if (sizes_provided == 0) {  // no clue how to resize - keep original size then
+    for (int d = 0; d < spatial_ndim_; d++)
+      out_size[d] = in_size[d];
+  } else {
+    if (mode_ == ResizeMode::Stretch) {
+      if (sizes_provided < spatial_ndim_) {
+        // if only some extents are provided - calculate average scale
+        // and use it for the missing dimensions;
+        double avg_scale = 1;
+        for (int d = 0; d < spatial_ndim_; d++) {
+          scale[d] = out_size[d] / in_size[d];
+          if (mask[d])
+            avg_scale *= std::abs(scale[d]);  // abs because of possible flipping
```

Comment:
Maybe we should limit the scale inference only to case when the scale is provided only for one dimension, or when there is more than one but the ratio is the same for each provided dimension.

Reply:
Well... it was more about not choosing the axis when the ratios are similar but not quite the same (for example due to rounding or unrepresentability of the ratio). BTW, I'm changing the meaning of "stretch" in upcoming commit (and renaming current behavior to "default") - the new "stretch" will not resize the unspecified dimensions at all.